src/libraries/System.Security.Cryptography/tests/X509Certificates/X500DistinguishedNameBuilderTests.cs

```diff
@@ -0,0 +1,379 @@
+        {
+            X500DistinguishedNameBuilder builder = new();
+            AssertExtensions.Throws<ArgumentNullException>(
+                "oid",
+                () => builder.AddEncoded((Oid)null, Array.Empty<byte>()));
+            AssertExtensions.Throws<ArgumentException>(
+                "oid",
+                () => builder.AddEncoded(new Oid(), Array.Empty<byte>()));
+            AssertExtensions.Throws<ArgumentException>(
+                "oid",
+                () => builder.AddEncoded(new Oid("", null), Array.Empty<byte>()));
+        }
+
+        [Fact]
+        public static void AddEncoded_Value_Null_Fails()
+        {
+            X500DistinguishedNameBuilder builder = new();
+            AssertExtensions.Throws<ArgumentNullException>(
+                "encodedValue",
+                () => builder.AddEncoded(TestOid, (byte[])null));
+            AssertExtensions.Throws<ArgumentNullException>(
+                "encodedValue",
+                () => builder.AddEncoded(new Oid(TestOid, null), (byte[])null));
+        }
+
+        [Fact]
+        public static void Add_InvalidUniversalTagNumber()
+        {
+            X500DistinguishedNameBuilder builder = new();
+            AssertExtensions.Throws<ArgumentException>(
+                "stringEncodingType",
+                () => builder.Add(TestOid, "True", UniversalTagNumber.Boolean));
+            AssertExtensions.Throws<ArgumentException>(
+                "stringEncodingType",
+                () => builder.Add(new Oid(TestOid, null), "True", UniversalTagNumber.Boolean));
+        }
+
+        [Fact]
+        public static void Add_OidString_NullOrEmpty_Fails()
+        {
+            X500DistinguishedNameBuilder builder = new();
+            AssertExtensions.Throws<ArgumentNullException>(
+                "oidValue",
+                () => builder.Add((string)null, "banana"));
+            AssertExtensions.Throws<ArgumentException>(
+                "oidValue",
+                () => builder.Add("", "banana"));
+        }
+
+        [Fact]
+        public static void Add_OidString_NotAnOid_Fails()
+        {
+            X500DistinguishedNameBuilder builder = new();
+            AssertExtensions.Throws<ArgumentException>(
+                "oidValue",
+                () => builder.Add("strawberry", "banana"));
+        }
+
+        [Theory]
+        [MemberData(nameof(AddStringTheories))]
+        public static void Add_StringOid_Success(string oid, string value, UniversalTagNumber? tag, string expectedHex)
+        {
+            AssertBuilder(expectedHex, builder => builder.Add(oid, value, tag));
+        }
+
+        [Theory]
+        [MemberData(nameof(AddStringTheories))]
+        public static void Add_Oid_Success(string oid, string value, UniversalTagNumber? tag, string expectedHex)
+        {
+            AssertBuilder(expectedHex, builder => builder.Add(new Oid(oid, null), value, tag));
+        }
+
```

Comment:
A lifetime test is probably also useful.

* Add some stuff, call Build.
* Add more stuff.  Note how it doesn't throw.  Call Build.  See that it was cumulative.